### PR TITLE
[wip] Add caching to local execution

### PR DESCRIPTION
**Pull request by eapolinario**
# TL;DR
Enable the ability to cache the output of tasks in local executions

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
[`joblib.Memory`](https://joblib.readthedocs.io/en/latest/generated/joblib.Memory.html#joblib.Memory) provides a persistent store to cache the result of function invocations. Similarly to how we do in the hosted case, we use the pair `(task inputs, cache version)` to memoize task outputs across different executions.

We're also exposing a configuration knob to let users decide where the cached objects will be written to disk.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/761

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/core/base_task.py [hunk 1]
```diff
@@ -236,7 +250,13 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
         )
         input_literal_map = _literal_models.LiteralMap(literals=kwargs)
 
-        outputs_literal_map = self.dispatch_execute(ctx, input_literal_map)
+        # TODO: improve comment
+        # if metadata.cache is set, check memoized version including cache_version
```
Comment (kumare3) [inline flytekit/core/base_task.py:254]: cc @wild-endeavor 
FYI, Yee is working on some updates to this part of the code as well

Comment (kumare3) [reply flytekit/core/base_task.py:254]: nope

File: Makefile [hunk 2]
```diff
@@ -88,3 +88,8 @@ update_version:
 
 	grep "$(PLACEHOLDER)" "setup.py"
 	sed -i "s/$(PLACEHOLDER)/__version__ = \"${VERSION}\"/g" "setup.py"
+
+# TODO
+.PHONY: clear-cache
+clear-cache:
```
Comment (kumare3) [inline Makefile:94]: can we make this a pyflyte command?

`pyflyte cache clear`

Comment (eapolinario) [reply Makefile:94]: good call, much cleaner.

**Review by kumare3 — APPROVED** [review]

File: flytekit/core/base_task.py [hunk 3]
```diff
@@ -236,7 +245,17 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
         )
         input_literal_map = _literal_models.LiteralMap(literals=kwargs)
 
-        outputs_literal_map = self.dispatch_execute(ctx, input_literal_map)
+        # if metadata.cache is set, check memoized version
+        if self._metadata.cache:
+            # The cache key is composed only of '(input_literal_map, cache_version)', i.e. all other parameters
```
Comment (katrogan) [inline flytekit/core/base_task.py:250]: should we also include the task name as part of the key?

similar to what we do with the identifier on hosted flyte https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/pluginmachinery/catalog/client.go#L26

File: tests/flytekit/unit/core/test_local_cache.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+from pytest import fixture
```
Comment (katrogan) [inline tests/flytekit/unit/core/test_local_cache.py:1]: can we test some more complex input and output types, like dataclasses, schemas, files?

File: flytekit/core/local_cache.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+CACHE_LOCATION = "~/.flyte/local-cache"
+
+
+class LocalCache(object):
+    _memory: Memory
+    _initialized: bool = False
+
+    @staticmethod
+    def initialize():
+        LocalCache._memory = Memory(CACHE_LOCATION, verbose=5)
```
Comment (katrogan) [inline flytekit/core/local_cache.py:16]: should we make the verbosity configurable too?

**Review by katrogan — COMMENTED** [review]
looks great, can we add some more tests!

File: flytekit/core/base_task.py (hunk 3 above)
Comment (eapolinario) [reply flytekit/core/base_task.py:250]: The key is already implicitly part of the cache key (in other words, `joblib.Memory` only caches the result of _each_ function call). I'm going to update the comment to reflect that.

File: flytekit/core/local_cache.py (hunk 5 above)
Comment (eapolinario) [reply flytekit/core/local_cache.py:16]: I don't think we gain a lot from making this configurable, especially given that configuration files are not used at all in local executions. I'm going to leave a TODO as well.

File: flytekit/core/base_task.py (hunk 3 above)
Comment (eapolinario) [reply flytekit/core/base_task.py:250]: Actually, this is a really good call. I'm going to add a test case to demonstrate the problem and subsequent fix.

Comment (eapolinario) [reply flytekit/core/base_task.py]: Fixed by https://github.com/flyteorg/flytekit/pull/592/commits/130b4327c7e9b5bad40626c0bdc52669270d97b9.

File: tests/flytekit/unit/core/test_local_cache.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        # The second and third calls hit the cache
+        assert (my_wf().open().all() == pandas.DataFrame(data={"x": [1, 2], "y": ["3", "4"]})).all().all()
+        assert n_cached_task_calls == 1
+        assert (my_wf().open().all() == pandas.DataFrame(data={"x": [1, 2], "y": ["3", "4"]})).all().all()
+        assert n_cached_task_calls == 1
+
+
+def test_wf_custom_types():
+    @dataclass_json
+    @dataclass
```
Comment (katrogan) [inline tests/flytekit/unit/core/test_local_cache.py:157]: awesome, this works? ran into issues with dataclass serialization last time i tried this but maybe joblib has since worked around this. woohoo!

**Review by katrogan — APPROVED** [review]

File: tests/flytekit/unit/core/test_local_cache.py (hunk 6 above)
Comment (kumare3) [reply tests/flytekit/unit/core/test_local_cache.py:157]: question around this, are we storing `LiteralMap` or are we storing the native `python values`. Also what happens to file refrences, do we store a shallow copy or a deep copy?
